File: sidecar/src/services/llm-gateway.ts
export interface LLMMessage {
  role: 'user' | 'assistant' | 'system'
  content: string
}

export interface LLMRequest {
  messages: LLMMessage[]
  model: string
  maxTokens?: number
  stream: boolean
  systemPrompt?: string
  onTokenUsage?: (input: number, output: number, total: number) => void
}

interface LLMProvider {
  name: string
  streamChat(req: LLMRequest): AsyncGenerator<string>
  models(): string[]
}

class AnthropicProvider implements LLMProvider {
  name = 'anthropic'
  private apiKey: string

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  async *streamChat(req: LLMRequest): AsyncGenerator<string> {
    const response = await fetch('https://api.anthropic.com/v1/messages', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': this.apiKey,
        'anthropic-version': '2023-06-01',
      },
      body: JSON.stringify({
        model: req.model,
        max_tokens: req.maxTokens ?? 8192,
        system: req.systemPrompt,
        messages: req.messages,
        stream: true,
      }),
      signal: AbortSignal.timeout(30_000),
    })

    if (!response.ok) {
      throw new Error(`Anthropic API error: ${response.status} ${await response.text()}`)
    }

    const reader = response.body?.getReader()
    if (!reader) throw new Error('No response body')

    const decoder = new TextDecoder()
    let buffer = ''
    let inputTokens = 0
    let outputTokens = 0

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const data = line.slice(6)
        if (data === '[DONE]') return

        try {
          const parsed = JSON.parse(data)
          if (parsed.type === 'message_start' && parsed.message?.usage) {
            inputTokens = parsed.message.usage.input_tokens || 0
          }
          if (parsed.type === 'message_delta' && parsed.usage) {
            outputTokens = parsed.usage.output_tokens || 0
          }
          if (parsed.type === 'content_block_delta' && parsed.delta?.text) {
            yield parsed.delta.text
          }
        } catch {
          // skip
        }
      }
    }

    if (req.onTokenUsage && (inputTokens || outputTokens)) {
      req.onTokenUsage(inputTokens, outputTokens, inputTokens + outputTokens)
    }
  }

  models(): string[] {
    return ['claude-sonnet-4-20250514', 'claude-3-5-sonnet-20241022', 'claude-opus-4-20250514']
  }
}

class OpenAIProvider implements LLMProvider {
  name = 'openai'
  private apiKey: string
  private baseUrl: string

  constructor(apiKey: string, baseUrl = 'https://api.openai.com/v1') {
    this.apiKey = apiKey
    this.baseUrl = baseUrl
  }

  async *streamChat(req: LLMRequest): AsyncGenerator<string> {
    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: req.model,
        messages: [
          ...(req.systemPrompt ? [{ role: 'system', content: req.systemPrompt }] : []),
          ...req.messages,
        ],
        max_tokens: req.maxTokens ?? 8192,
        stream: true,
        stream_options: { include_usage: true },
      }),
      signal: AbortSignal.timeout(30_000),
    })

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.status} ${await response.text()}`)
    }

    const reader = response.body?.getReader()
    if (!reader) throw new Error('No response body')

    const decoder = new TextDecoder()
    let buffer = ''
    let inputTokens = 0
    let outputTokens = 0

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const data = line.slice(6)
        if (data === '[DONE]') return

        try {
          const parsed = JSON.parse(data)
          if (parsed.usage) {
            inputTokens = parsed.usage.prompt_tokens || 0
            outputTokens = parsed.usage.completion_tokens || 0
          }
          const content = parsed.choices?.[0]?.delta?.content
          if (content) yield content
        } catch {
          // skip
        }
      }
    }

    if (req.onTokenUsage && (inputTokens || outputTokens)) {
      req.onTokenUsage(inputTokens, outputTokens, inputTokens + outputTokens)
    }
  }

  models(): string[] {
    if (this.baseUrl.includes('groq')) {
      return [
        'llama3-70b-8192',
        'llama3-8b-8192',
        'mixtral-8x7b-32768',
        'gemma2-9b-it',
        'gemma-7b-it',
      ]
    }
    return ['gpt-4o', 'gpt-4o-mini', 'o1', 'o3-mini']
  }
}

export class LLMGateway {
  private providers: Map<string, LLMProvider> = new Map()
  private fallbackOrder: string[] = []

  registerProvider(provider: LLMProvider): void {
    this.providers.set(provider.name, provider)
  }

  setFallbackOrder(order: string[]): void {
    this.fallbackOrder = order
  }

  private extractRetryDelay(errorMessage: string): number | null {
    const match = errorMessage.match(/retry[_\s]?delay["']?:\s*"?(\d+(?:\.\d+)?)\s*s/i)
    if (match) return parseFloat(match[1]) * 1000
    const retryAfter = errorMessage.match(/retry after (\d+)/i)
    if (retryAfter) return parseFloat(retryAfter[1]) * 1000
    return null
  }

  async *stream(req: LLMRequest): AsyncGenerator<string> {
    const provider = this.getProviderForModel(req.model)
    if (!provider) {
      throw new Error(`No provider found for model: ${req.model}`)
    }

    const errors: Error[] = []

    try {
      yield* provider.streamChat(req)
    } catch (err) {
      const errMsg = (err as Error).message
      errors.push(err as Error)

      if (errMsg.includes('429') || errMsg.includes('RESOURCE_EXHAUSTED')) {
        const delay = this.extractRetryDelay(errMsg) ?? 15_000
        await new Promise(resolve => setTimeout(resolve, Math.min(delay, 30_000)))
        try {
          yield* provider.streamChat(req)
          return
        } catch (retryErr) {
          errors.push(retryErr as Error)
        }
      }

      for (const fallbackName of this.fallbackOrder) {
        const fallback = this.providers.get(fallbackName)
        if (!fallback) continue

        try {
          yield* fallback.streamChat(req)
          return
        } catch (fallbackErr) {
          errors.push(fallbackErr as Error)
        }
      }

      throw new Error(`All providers failed: ${errors.map(e => e.message).join(', ')}`)
    }
  }

  private getProviderForModel(model: string): LLMProvider | undefined {
    for (const provider of this.providers.values()) {
      if (provider.models().includes(model)) {
        return provider
      }
    }

    if (model.startsWith('claude-')) return this.providers.get('anthropic')
    if (model.startsWith('gpt-') || model.startsWith('o1') || model.startsWith('o3')) {
      return this.providers.get('openai')
    }
    if (model.startsWith('gemini-')) return this.providers.get('gemini')
    if (model.startsWith('deepseek-')) return this.providers.get('deepseek')
    if (model.includes('mistral') || model.includes('codestral')) return this.providers.get('mistral')
    if (model.includes('llama') || model.includes('mixtral') || model.includes('gemma')) {
      return this.providers.get('groq')
    }
    if (model.startsWith('command-')) return this.providers.get('cohere')

    return this.providers.get('groq') || this.providers.get('openai')
  }

  getModels(): string[] {
    const models: string[] = []
    for (const provider of this.providers.values()) {
      models.push(...provider.models())
    }
    return [...new Set(models)]
  }
}

class GeminiProvider implements LLMProvider {
  name = 'gemini'
  private apiKey: string

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  async *streamChat(req: LLMRequest): AsyncGenerator<string> {
    const response = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/${req.model}:streamGenerateContent?key=${this.apiKey}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: req.messages.map(m => ({
          role: m.role === 'assistant' ? 'model' : 'user',
          parts: [{ text: m.content }],
        })),
        systemInstruction: req.systemPrompt ? {
          parts: [{ text: req.systemPrompt }],
        } : undefined,
        generationConfig: {
          maxOutputTokens: req.maxTokens ?? 8192,
        },
      }),
      signal: AbortSignal.timeout(30_000),
    })

    if (!response.ok) {
      throw new Error(`Gemini API error: ${response.status} ${await response.text()}`)
    }

    const reader = response.body?.getReader()
    if (!reader) throw new Error('No response body')

    const decoder = new TextDecoder()
    let buffer = ''
    let inputTokens = 0
    let outputTokens = 0

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const data = line.slice(6)
        if (data === '[DONE]') return

        try {
          const parsed = JSON.parse(data)
          if (parsed.usageMetadata) {
            inputTokens = parsed.usageMetadata.promptTokenCount || 0
            outputTokens = parsed.usageMetadata.candidatesTokenCount || 0
          }
          const text = parsed.candidates?.[0]?.content?.parts?.[0]?.text
          if (text) yield text
        } catch {
          // skip
        }
      }
    }

    if (req.onTokenUsage && (inputTokens || outputTokens)) {
      req.onTokenUsage(inputTokens, outputTokens, inputTokens + outputTokens)
    }
  }

  models(): string[] {
    return ['gemini-2.0-flash', 'gemini-2.0-flash-lite', 'gemini-2.5-pro-exp-03-25']
  }
}

class DeepSeekProvider extends OpenAIProvider {
  declare name: string

  constructor(apiKey: string) {
    super(apiKey, 'https://api.deepseek.com/v1')
    this.name = 'deepseek'
  }

  models(): string[] {
    return ['deepseek-chat', 'deepseek-reasoner']
  }
}

class MistralProvider extends OpenAIProvider {
  declare name: string

  constructor(apiKey: string) {
    super(apiKey, 'https://api.mistral.ai/v1')
    this.name = 'mistral'
  }

  models(): string[] {
    return ['mistral-large-latest', 'codestral-latest', 'mistral-small-latest']
  }
}

class CohereProvider implements LLMProvider {
  name = 'cohere'
  private apiKey: string

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  async *streamChat(req: LLMRequest): AsyncGenerator<string> {
    const response = await fetch('https://api.cohere.com/v2/chat', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: req.model,
        messages: [
          ...(req.systemPrompt ? [{ role: 'system', content: req.systemPrompt }] : []),
          ...req.messages.map(m => ({ role: m.role, content: m.content })),
        ],
        max_tokens: req.maxTokens ?? 8192,
        stream: true,
      }),
      signal: AbortSignal.timeout(30_000),
    })

    if (!response.ok) {
      throw new Error(`Cohere API error: ${response.status} ${await response.text()}`)
    }

    const reader = response.body?.getReader()
    if (!reader) throw new Error('No response body')

    const decoder = new TextDecoder()
    let buffer = ''
    let inputTokens = 0
    let outputTokens = 0

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const data = line.slice(6)
        if (data === '[DONE]') return

        try {
          const parsed = JSON.parse(data)
          if (parsed.type === 'text-generation' && parsed.text) {
            yield parsed.text
          }
          if (parsed.type === 'stream-end') {
            inputTokens = parsed.response?.meta?.billed_units?.input_tokens || 0
            outputTokens = parsed.response?.meta?.billed_units?.output_tokens || 0
          }
        } catch {
          // skip
        }
      }
    }

    if (req.onTokenUsage && (inputTokens || outputTokens)) {
      req.onTokenUsage(inputTokens, outputTokens, inputTokens + outputTokens)
    }
  }

  models(): string[] {
    return ['command-a-03-2025', 'command-r-08-2024', 'command-r-plus-08-2024', 'command-r7b-12-2024']
  }
}

let _currentGateway: LLMGateway | null = null

export function setCurrentGateway(g: LLMGateway) {
  _currentGateway = g
}

export function getCurrentGateway(): LLMGateway {
  if (!_currentGateway) throw new Error('Gateway not initialized')
  return _currentGateway
}

export function createGateway(apiKeys: Record<string, string>): LLMGateway {
  const gateway = new LLMGateway()

  if (apiKeys.anthropic) {
    gateway.registerProvider(new AnthropicProvider(apiKeys.anthropic))
  }
  if (apiKeys.openai) {
    gateway.registerProvider(new OpenAIProvider(apiKeys.openai))
  }
  if (apiKeys.openrouter) {
    gateway.registerProvider(new OpenAIProvider(apiKeys.openrouter, 'https://openrouter.ai/api/v1'))
  }
  if (apiKeys.groq) {
    gateway.registerProvider(new OpenAIProvider(apiKeys.groq, 'https://api.groq.com/openai/v1'))
  }
  if (apiKeys.gemini) {
    gateway.registerProvider(new GeminiProvider(apiKeys.gemini))
  }
  if (apiKeys.deepseek) {
    gateway.registerProvider(new DeepSeekProvider(apiKeys.deepseek))
  }
  if (apiKeys.mistral) {
    gateway.registerProvider(new MistralProvider(apiKeys.mistral))
  }
  if (apiKeys.cohere) {
    gateway.registerProvider(new CohereProvider(apiKeys.cohere))
  }

  gateway.setFallbackOrder(['anthropic', 'gemini', 'openai', 'deepseek', 'mistral', 'cohere', 'groq'])

  return gateway
}

File: src-tauri/src/commands/llm_gateway.rs
use aes_gcm::{
    aead::Aead,
    Aes256Gcm, Key, KeyInit, Nonce,
};
use base64::Engine;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use tauri::ipc::Channel;

use super::ai::ChatStreamEvent;

const CRYPTO_KEY_HEX: &str = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";

fn decrypt_key(ciphertext_b64: &str) -> Result<String, String> {
    let key_bytes = hex::decode(CRYPTO_KEY_HEX).map_err(|e| format!("Invalid crypto key hex: {}", e))?;
    let key = Key::<Aes256Gcm>::from_slice(&key_bytes);
    let cipher = Aes256Gcm::new(key);

    let data = base64::engine::general_purpose::STANDARD
        .decode(ciphertext_b64)
        .map_err(|e| format!("Invalid base64: {}", e))?;

    if data.len() < 32 {
        return Err("Ciphertext too short".into());
    }

    let (iv_bytes, rest) = data.split_at(16);
    let (tag_bytes, encrypted) = rest.split_at(16);

    let nonce = Nonce::from_slice(iv_bytes);
    let mut payload = encrypted.to_vec();
    payload.extend_from_slice(tag_bytes);

    let plaintext = cipher
        .decrypt(nonce, payload.as_ref())
        .map_err(|e| format!("Decryption failed: {:?}", e))?;

    String::from_utf8(plaintext).map_err(|e| format!("Invalid UTF-8: {}", e))
}

fn find_db_path() -> PathBuf {
    let candidates = [
        PathBuf::from("D:\\Projeto Fluxcodex\\ai-app-builder\\sidecar\\prisma\\aibuilder.db"),
        {
            let mut p = std::env::current_dir().unwrap_or_default();
            p.push("sidecar/prisma/aibuilder.db");
            p
        },
        {
            let mut p = std::env::current_dir().unwrap_or_default();
            p.push("prisma/aibuilder.db");
            p
        },
    ];

    for c in &candidates {
        if c.exists() {
            return c.clone();
        }
    }

    candidates[0].clone()
}

fn load_api_keys() -> Result<HashMap<String, String>, String> {
    let db_path = find_db_path();
    if !db_path.exists() {
        return Ok(HashMap::new());
    }

    let conn = rusqlite::Connection::open(&db_path).map_err(|e| format!("DB open: {}", e))?;
    let mut stmt = conn
        .prepare("SELECT provider, keyHash FROM ApiKey")
        .map_err(|e| format!("SQL prepare: {}", e))?;

    let rows = stmt
        .query_map([], |row| {
            let provider: String = row.get(0)?;
            let key_hash: String = row.get(1)?;
            Ok((provider, key_hash))
        })
        .map_err(|e| format!("SQL query: {}", e))?;

    let mut keys = HashMap::new();
    for row in rows {
        let (provider, key_hash) = row.map_err(|e| format!("Row read: {}", e))?;
        if let Ok(decrypted) = decrypt_key(&key_hash) {
            keys.insert(provider, decrypted);
        }
    }

    Ok(keys)
}

#[derive(Deserialize)]
pub struct CloudChatRequest {
    pub message: String,
    pub mode: Option<String>,
    pub model: String,
    pub active_skills: Option<Vec<super::ai::ActiveSkill>>,
}

#[derive(Serialize)]
pub struct ModelGroup {
    pub provider: String,
    pub models: Vec<String>,
}

fn build_system_prompt(mode: &str, active_skills: &[super::ai::ActiveSkill]) -> String {
    let mut prompt = String::from(
        "# Identidade\nVocê é o AI App Builder Studio — um assistente nativo do Fluxcodex. Responda em português brasileiro.\n",
    );

    if mode == "think" {
        prompt.push_str(
            "\n## Modo Think\nAntes de responder, organize a solução em etapas curtas, verifique riscos e entregue a resposta final de forma objetiva.\n",
        );
    }

    if !active_skills.is_empty() {
        prompt.push_str("\n## Skills Ativas\n");
        for skill in active_skills {
            prompt.push_str(&format!("### {}\n{}\n{}\n", skill.name, skill.description, skill.content));
        }
    }

    prompt
}

fn get_provider_for_model(model: &str) -> &str {
    if model.starts_with("claude-") {
        "anthropic"
    } else if model.starts_with("gpt-") || model.starts_with("o1") || model.starts_with("o3") {
        "openai"
    } else if model.starts_with("gemini-") {
        "gemini"
    } else if model.starts_with("deepseek-") {
        "deepseek"
    } else if model.contains("mistral") || model.contains("codestral") {
        "mistral"
    } else if model.contains("llama") || model.contains("mixtral") || model.contains("gemma") {
        "groq"
    } else {
        "openai"
    }
}

fn get_provider_base_url(provider: &str) -> &str {
    match provider {
        "deepseek" => "https://api.deepseek.com/v1",
        "mistral" => "https://api.mistral.ai/v1",
        "groq" => "https://api.groq.com/openai/v1",
        "openrouter" => "https://openrouter.ai/api/v1",
        _ => "https://api.openai.com/v1",
    }
}

async fn stream_anthropic(
    api_key: &str,
    model: &str,
    message: &str,
    system_prompt: &str,
    on_event: &Channel<ChatStreamEvent>,
) -> Result<(), String> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(120))
        .build()
        .map_err(|e| format!("Client build: {}", e))?;
    let mut response = client
        .post("https://api.anthropic.com/v1/messages")
        .header("Content-Type", "application/json")
        .header("x-api-key", api_key)
        .header("anthropic-version", "2023-06-01")
        .json(&serde_json::json!({
            "model": model,
            "max_tokens": 8192,
            "system": system_prompt,
            "messages": [{"role": "user", "content": message}],
            "stream": true,
        }))
        .send()
        .await
        .map_err(|e| format!("Anthropic request: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(format!("Anthropic API {}: {}", status, body));
    }

    super::ai::reset_cancel();

    let mut buffer = String::new();
    let mut input_tokens = 0u64;
    let mut output_tokens = 0u64;

    loop {
        if super::ai::is_cancelled() {
            return Ok(());
        }

        let chunk = response
            .chunk()
            .await
            .map_err(|e| format!("Anthropic stream: {}", e))?;
        let chunk = match chunk {
            Some(c) => c,
            None => break,
        };

        if let Ok(text) = String::from_utf8(chunk.to_vec()) {
            buffer.push_str(&text);
            while let Some(pos) = buffer.find('\n') {
                let line = buffer[..pos].trim().to_string();
                buffer = buffer[pos + 1..].to_string();

                if !line.starts_with("data: ") {
                    continue;
                }
                let data = line[6..].trim();
                if data == "[DONE]" {
                    break;
                }

                if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(msg) = parsed.get("message") {
                        if let Some(usage) = msg.get("usage") {
                            input_tokens = usage.get("input_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                        }
                    }
                    if parsed.get("type").and_then(|v| v.as_str()) == Some("message_delta") {
                        if let Some(usage) = parsed.get("usage") {
                            output_tokens = usage.get("output_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                        }
                    }
                    if let Some(delta) = parsed.get("delta") {
                        if let Some(text) = delta.get("text").and_then(|v| v.as_str()) {
                            if !text.is_empty() {
                                on_event
                                    .send(ChatStreamEvent::Chunk {
                                        content: text.to_string(),
                                    })
                                    .map_err(|e| e.to_string())?;
                            }
                        }
                    }
                }
            }
        }
    }

    if input_tokens > 0 || output_tokens > 0 {
        on_event
            .send(ChatStreamEvent::TokenUsage {
                total: (input_tokens + output_tokens) as u32,
            })
            .ok();
    }

    Ok(())
}

async fn stream_openai_compatible(
    api_key: &str,
    base_url: &str,
    model: &str,
    message: &str,
    system_prompt: &str,
    on_event: &Channel<ChatStreamEvent>,
) -> Result<(), String> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(120))
        .build()
        .map_err(|e| format!("Client build: {}", e))?;
    let mut response = client
        .post(format!("{}/chat/completions", base_url))
        .header("Content-Type", "application/json")
        .header("Authorization", format!("Bearer {}", api_key))
        .json(&serde_json::json!({
            "model": model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": message}
            ],
            "max_tokens": 8192,
            "stream": true,
            "stream_options": {"include_usage": true},
        }))
        .send()
        .await
        .map_err(|e| format!("OpenAI request: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(format!("OpenAI API {}: {}", status, body));
    }

    super::ai::reset_cancel();

    let mut buffer = String::new();
    let mut input_tokens = 0u64;
    let mut output_tokens = 0u64;

    loop {
        if super::ai::is_cancelled() {
            return Ok(());
        }

        let chunk = response
            .chunk()
            .await
            .map_err(|e| format!("OpenAI stream: {}", e))?;
        let chunk = match chunk {
            Some(c) => c,
            None => break,
        };

        if let Ok(text) = String::from_utf8(chunk.to_vec()) {
            buffer.push_str(&text);
            while let Some(pos) = buffer.find('\n') {
                let line = buffer[..pos].trim().to_string();
                buffer = buffer[pos + 1..].to_string();

                if !line.starts_with("data: ") {
                    continue;
                }
                let data = line[6..].trim();
                if data == "[DONE]" {
                    break;
                }

                if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(usage) = parsed.get("usage") {
                        input_tokens = usage.get("prompt_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                        output_tokens = usage.get("completion_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                    }
                    if let Some(choices) = parsed.get("choices").and_then(|v| v.as_array()) {
                        if let Some(choice) = choices.first() {
                            if let Some(delta) = choice.get("delta") {
                                if let Some(content) = delta.get("content").and_then(|v| v.as_str()) {
                                    if !content.is_empty() {
                                        on_event
                                            .send(ChatStreamEvent::Chunk {
                                                content: content.to_string(),
                                            })
                                            .map_err(|e| e.to_string())?;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if input_tokens > 0 || output_tokens > 0 {
        on_event
            .send(ChatStreamEvent::TokenUsage {
                total: (input_tokens + output_tokens) as u32,
            })
            .ok();
    }

    Ok(())
}

async fn stream_gemini(
    api_key: &str,
    model: &str,
    message: &str,
    system_prompt: &str,
    on_event: &Channel<ChatStreamEvent>,
) -> Result<(), String> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(120))
        .build()
        .map_err(|e| format!("Client build: {}", e))?;
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:streamGenerateContent?key={}",
        model, api_key
    );

    let mut response = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&serde_json::json!({
            "contents": [{"role": "user", "parts": [{"text": message}]}],
            "systemInstruction": {"parts": [{"text": system_prompt}]},
            "generationConfig": {"maxOutputTokens": 8192},
        }))
        .send()
        .await
        .map_err(|e| format!("Gemini request: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(format!("Gemini API {}: {}", status, body));
    }

    super::ai::reset_cancel();

    let mut buffer = String::new();
    let mut input_tokens = 0u64;
    let mut output_tokens = 0u64;

    loop {
        if super::ai::is_cancelled() {
            return Ok(());
        }

        let chunk = response
            .chunk()
            .await
            .map_err(|e| format!("Gemini stream: {}", e))?;
        let chunk = match chunk {
            Some(c) => c,
            None => break,
        };

        if let Ok(text) = String::from_utf8(chunk.to_vec()) {
            buffer.push_str(&text);
            while let Some(pos) = buffer.find('\n') {
                let line = buffer[..pos].trim().to_string();
                buffer = buffer[pos + 1..].to_string();

                if !line.starts_with("data: ") {
                    continue;
                }
                let data = line[6..].trim();
                if data == "[DONE]" {
                    break;
                }

                if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(meta) = parsed.get("usageMetadata") {
                        input_tokens = meta.get("promptTokenCount").and_then(|v| v.as_u64()).unwrap_or(0);
                        output_tokens = meta.get("candidatesTokenCount").and_then(|v| v.as_u64()).unwrap_or(0);
                    }
                    if let Some(candidates) = parsed.get("candidates").and_then(|v| v.as_array()) {
                        if let Some(candidate) = candidates.first() {
                            if let Some(content) = candidate.get("content") {
                                if let Some(parts) = content.get("parts").and_then(|v| v.as_array()) {
                                    for part in parts {
                                        if let Some(text) = part.get("text").and_then(|v| v.as_str()) {
                                            if !text.is_empty() {
                                                on_event
                                                    .send(ChatStreamEvent::Chunk {
                                                        content: text.to_string(),
                                                    })
                                                    .map_err(|e| e.to_string())?;
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if input_tokens > 0 || output_tokens > 0 {
        on_event
            .send(ChatStreamEvent::TokenUsage {
                total: (input_tokens + output_tokens) as u32,
            })
            .ok();
    }

    Ok(())
}

#[tauri::command]
pub async fn cloud_chat_completion(
    request: CloudChatRequest,
    on_event: Channel<ChatStreamEvent>,
) -> Result<(), String> {
    super::ai::reset_cancel();

    let mode = request.mode.unwrap_or_else(|| "chat".into());
    let model = request.model;
    let active_skills = request.active_skills.unwrap_or_default();
    let system_prompt = build_system_prompt(&mode, &active_skills);

    let provider_name = get_provider_for_model(&model);
    let keys = load_api_keys()?;

    let api_key = keys.get(provider_name).ok_or_else(|| {
        format!(
            "API key não encontrada para '{}'. Adicione em Configurações > API Keys.",
            provider_name
        )
    })?;

    if super::ai::is_cancelled() {
        return Ok(());
    }

    match provider_name {
        "anthropic" => stream_anthropic(api_key, &model, &request.message, &system_prompt, &on_event).await,
        "gemini" => stream_gemini(api_key, &model, &request.message, &system_prompt, &on_event).await,
        _ => {
            let base_url = get_provider_base_url(provider_name);
            stream_openai_compatible(api_key, base_url, &model, &request.message, &system_prompt, &on_event).await
        }
    }
}

#[tauri::command]
pub fn cloud_models() -> Vec<ModelGroup> {
    vec![
        ModelGroup {
            provider: "Anthropic".into(),
            models: vec![
                "claude-sonnet-4-20250514".into(),
                "claude-3-5-sonnet-20241022".into(),
                "claude-opus-4-20250514".into(),
            ],
        },
        ModelGroup {
            provider: "OpenAI".into(),
            models: vec!["gpt-4o".into(), "gpt-4o-mini".into(), "o1".into(), "o3-mini".into()],
        },
        ModelGroup {
            provider: "Gemini".into(),
            models: vec![
                "gemini-2.0-flash".into(),
                "gemini-2.0-flash-lite".into(),
                "gemini-2.5-pro-exp-03-25".into(),
            ],
        },
        ModelGroup {
            provider: "DeepSeek".into(),
            models: vec!["deepseek-chat".into(), "deepseek-reasoner".into()],
        },
        ModelGroup {
            provider: "Mistral".into(),
            models: vec![
                "mistral-large-latest".into(),
                "codestral-latest".into(),
                "mistral-small-latest".into(),
            ],
        },
        ModelGroup {
            provider: "Groq".into(),
            models: vec![
                "llama3-70b-8192".into(),
                "llama3-8b-8192".into(),
                "mixtral-8x7b-32768".into(),
                "gemma2-9b-it".into(),
            ],
        },
    ]
}
